File: app/backend/api/src/main/java/com/example/sms/domain/model/master/partner/Vendor.java
package com.example.sms.domain.model.master.partner;

import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Value;

/**
 * 仕入先
 */
@Value
@AllArgsConstructor
@NoArgsConstructor(force = true)
public class Vendor {
    VendorCode vendorCode; // 仕入先コード
    VendorName vendorName; // 仕入先名
    String vendorContactName; // 仕入先担当者名
    String vendorDepartmentName; // 仕入先部門名
    Address vendorAddress; // 仕入先住所
    String vendorPhoneNumber; // 仕入先電話番号
    String vendorFaxNumber; // 仕入先ｆａｘ番号
    String vendorEmailAddress; // 仕入先メールアドレス
    Integer vendorClosingDate; // 仕入先締日
    Integer vendorPaymentMonth; // 仕入先支払月
    Integer vendorPaymentDate; // 仕入先支払日
    Integer vendorPaymentMethod; // 仕入先支払方法

    public static Vendor of(
            String vendorCode,
            Integer vendorBranchCode,
            String vendorName,
            String vendorNameKana,
            String vendorContactName,
            String vendorDepartmentName,
            String vendorPostalCode,
            String vendorPrefecture,
            String vendorAddress1,
            String vendorAddress2,
            String vendorPhoneNumber,
            String vendorFaxNumber,
            String vendorEmailAddress,
            Integer vendorClosingDate,
            Integer vendorPaymentMonth,
            Integer vendorPaymentDate,
            Integer vendorPaymentMethod
    ) {
        return new Vendor(
                VendorCode.of(vendorCode, vendorBranchCode),
                VendorName.of(vendorName, vendorNameKana),
                vendorContactName,
                vendorDepartmentName,
                Address.of(vendorPostalCode, vendorPrefecture, vendorAddress1, vendorAddress2),
                vendorPhoneNumber,
                vendorFaxNumber,
                vendorEmailAddress,
                vendorClosingDate,
                vendorPaymentMonth,
                vendorPaymentDate,
                vendorPaymentMethod
        );
    }
}

File: app/frontend/src/components/master/partner/hooks/vendor.ts
import {useState} from "react";
import {
    VendorCriteriaType,
    VendorType
} from "../../../../models/master/partner";
import {useFetchEntities} from "../../../application/hooks.ts";
import {PageNationType} from "../../../../views/application/PageNation.tsx";
import {VendorService} from "../../../../services/master/vendor.ts";
import {
    ClosingDateEnumType,
    PaymentDayEnumType,
    PaymentMethodEnumType,
    PaymentMonthEnumType,
    PrefectureEnumType
} from "../../../../models";

export const useVendor = () => {
    const initialVendor: VendorType = {
        vendorCode: "",
        vendorBranchNumber: 0,
        vendorName: "",
        vendorNameKana: "",
        vendorContactName: "",
        vendorDepartmentName: "",
        vendorPostalCode: "",
        vendorPrefecture: PrefectureEnumType.東京都,
        vendorAddress1: "",
        vendorAddress2: "",
        vendorPhoneNumber: "",
        vendorFaxNumber: "",
        vendorEmailAddress: "",
        vendorClosingDate: ClosingDateEnumType.末日,
        vendorPaymentMonth: PaymentMonthEnumType.当月,
        vendorPaymentDate: PaymentDayEnumType.末日,
        vendorPaymentMethod: PaymentMethodEnumType.振込,
        checked: false
    };

    const [vendors, setVendors] = useState<VendorType[]>([]);
    const [newVendor, setNewVendor] = useState<VendorType>(initialVendor);
    const [searchVendorCriteria, setSearchVendorCriteria] = useState<VendorCriteriaType>({});
    const vendorService = VendorService();

    return {
        initialVendor,
        vendors,
        newVendor,
        setNewVendor,
        searchVendorCriteria,
        setSearchVendorCriteria,
        setVendors,
        vendorService,
    };
};

export const useFetchVendors = (
    setLoading: (loading: boolean) => void,
    setList: (list: VendorType[]) => void,
    setPageNation: (pageNation: PageNationType) => void,
    setError: (error: string) => void,
    showErrorMessage: (message: string, callback: (error: string) => void) => void,
    service: ReturnType<typeof VendorService>
) => {
    return useFetchEntities<VendorType, typeof service, VendorCriteriaType>(
        setLoading,
        setList,
        setPageNation,
        setError,
        showErrorMessage,
        service,
        "仕入先情報の取得に失敗しました:"
    );
};
